### ISSUE TITLE
Make etccalico mount path configurable

### ISSUE THREAD

**Issue by laibe** [tigera/operator#2444]
## Current Behavior

Currently the path is hardcoded to `/etc/calico` here: https://github.com/tigera/operator/blob/673ecc4c4d9774dcab03942ea93fb212a26e5cb4/pkg/render/csi.go#L140 

This is a problem for Talos OS that allows for writes only under `/var`.

## Possible Solution

make path configurable, e.g. so we can set it to `/var/calico` 

## Context

I would like to deploy a talos cluster with calico

## Your Environment

Talos 1.3

**Comment by epleterte** [comment]
Hi,
I'd like to see the same! I am also on Talos.

I don't really need the mount path to be configurable, I need the hostPath to be configurable.
I think it's these lines right here (currently starting on line 216):


		{
			Name: "etccalico",
			VolumeSource: corev1.VolumeSource{
				HostPath: &corev1.HostPathVolumeSource{
					Path: filepath.Clean("/etc/calico"),
				},
			},

My current test/workaround has been this:

Edit csi-node-driver DaemonSet and set this annotation:

`unsupported.operator.tigera.io/ignore: "true"`

...to make the operator leave the DS alone and not reset our next change, and then to change hostPath /etc/calico in csi-node-driver DS to something else:

      - hostPath:
          path: /var/lib/calico/etc
          type: ""
        name: etccalico

This works, but I have to leave the DS 'tainted' to keep it running.

**Comment by uhthomas** [comment]
@caseydavenport Apologies for the ping, but I believe this issue blocks anyone using an OS which does not persist `/etc/` from using Tigera. Unfortunately I will have to use Cilium in the meantime.

**Comment by tmjd** [comment]
@laibe do you think you need the MountPath configurable, I think it seems more reasonable that you would need the hostPath configurable like @epleterte suggested?

@epleterte using the unsupported annotation will be an issue when you upgrade as that component won't be upgraded with the annotation.

@uhthomas are you suggesting there are OS's that will discard `/etc` while they are running? I'm asking because I would expect a node that was rebooted to have the CSI driver pod redeployed so no need for `/etc` to be persisted across reboots. (When I hear about something not persisting it is usually in reference to across reboots but maybe you have a unique case.) Also incase you are unaware, the CSI driver I believe is only needed when using the Istio integration, so if you are not using that then you can disable it by setting the `KubeletVolumePluginPath` to `None`.

@Josh-Tigera it looks like you originally added CSI to the operator, so I assume you know a decent amount about the CSI driver and its installation, would there be any concern about exposing a field to set a different path instead of `/etc/calico`? Would there be any other changes necessary besides creating a field and then setting the Volume hostPath?

**Comment by Josh-Tigera** [comment]
> 

I seem to remember there being some paths (possibly /etc/calico but not exactly sure) that needed to be kept in sync with the node-driver-registrar so in addition to the field it may or may not be necessary to propagate the new path through to the image init call. The good news is that things will explode fairly quickly if you get it wrong so it won't be easy to miss.

**Comment by uhthomas** [comment]
@tmjd I believe the issue is that on some operating systes, like Talos, `/etc/` is read-only which means Calico cannot write to that path at all.

**Comment by epleterte** [comment]
Setting the annotation to keep the operator away was just a test to see that everything starts as expected.

@uhthomas You can skip using the CSI driver altogether if you don't need it by leaving KubeletVolumePluginPath unset. CSI is needed for Application Layer Policy, required i.e. for Istio, but is not required to use Calico.

**Comment by Josh-Tigera** [comment]
> Setting the annotation to keep the operator away was just a test to see that everything starts as expected.
> 
> @uhthomas You can skip using the CSI driver altogether if you don't need it by leaving KubeletVolumePluginPath unset. CSI is needed for Application Layer Policy, required i.e. for Istio, but is not required to use Calico.

@epleterte @uhthomas you actually need to explicitly set it to `None` if you want CSI disabled rather than leave it unset since unset will cause it to take on the default value as documented here: https://unified-docs.tigera.io/calico/3.25/reference/installation/api#operator.tigera.io/v1.Installation

**Comment by uhthomas** [comment]
Would you be able to clarify what the Calico CSI driver does? I am using [rook](https://rook.io/), would it affect that?

**Comment by Josh-Tigera** [comment]
> Would you be able to clarify what the Calico CSI driver does? I am using [rook](https://rook.io/), would it affect that?

@uhthomas the Calico CSI driver merely enables standard k8s CSI functionality so that arbitrary storage interfaces can be used. You can read more about it here: https://kubernetes.io/docs/concepts/storage/volumes/#csi

**Comment by uhthomas** [comment]
> > Would you be able to clarify what the Calico CSI driver does? I am using [rook](https://rook.io/), would it affect that?
> 
> 
> 
> @uhthomas the Calico CSI driver merely enables standard k8s CSI functionality so that arbitrary storage interfaces can be used. You can read more about it here: https://kubernetes.io/docs/concepts/storage/volumes/#csi

Ah :/ So definitely not compatible with rook then. I guess in that case I and many others are definitely still blocked on this.

**Comment by uhthomas** [comment]
@caseydavenport @Josh-Tigera I do have a proposed change here: #2518. I'd appreciate your feedback.

**Comment by Josh-Tigera** [comment]
> @uhthomas this shouldn't have any interference with your Rook installation. This is just a way of Calico to write some configuration to the host using a custom storage implementation. It can also be disabled if you're not using application layer policies as mentioned above.
> 
> I think we should make the etc/calico hostpath configurable for the CSI driver though - @Josh-Tigera it should be pretty straightforward no?
> 
> This is the only place it is used:
> 
> https://github.com/tigera/operator/blob/03a6123c3ba032ae85889532b4fca90792954634/pkg/render/csi.go#L200-L207
> 
> In the meantime, if you're not using ALP, you can just turn off the CSI driver altogether by setting the following config:
> 
> ```
> spec:
>   kubeletVolumePluginPath: None
> ```

If it's exclusively for CSI then yeah the changes are very straightforward. @uhthomas looked at your change it looks simple enough. Would be good to run through the steps Matt mentioned on your PR. I approved the workflow so we can at least see how the tests look. If something with that path goes awry it should be pretty obvious as the CSI init should fail loudly.

**Comment by ErikLundJensen** [comment]
We have tested the pull-request at an immutable Kubernetes installation (Sidero Talos 1.3.5)
Please go forward with this fix as it is really needed.

**Comment by tmjd** [comment]
For anyone following along here but not part of the conversation on https://github.com/tigera/operator/pull/2518, I've suggested that the `/etc/calico` volume be removed. Check [my message](https://github.com/tigera/operator/pull/2518#issuecomment-1551839167) for more details.

**Comment by uhthomas** [comment]
@tmjd Please see #2654.
